rusk/src/lib/services/prover/handler.rs

```diff
@@ -0,0 +1,153 @@
+                input.opening().clone(),
+                *input.note(),
+                input.pk_r_prime().into(),
+                input.value(),
+                input.blinding_factor(),
+                input.nullifier(),
+                cis,
+            );
+            circ.add_input(cinput).map_err(|e| {
+                Status::internal(format!(
+                    "Failed adding input to circuit: {}",
+                    e
+                ))
+            })?;
+        }
+        for (note, value, blinder) in utx.outputs() {
+            circ.add_output_with_data(*note, *value, *blinder).map_err(
+                |e| {
+                    Status::internal(format!(
+                        "Failed adding output to circuit: {}",
+                        e
+                    ))
+                },
+            )?;
+        }
+
+        circ.set_tx_hash(utx.hash());
+        circ.set_fee(utx.fee()).map_err(|e| {
+            Status::invalid_argument(format!("Failed setting fee: {}", e))
+        })?;
+
+        let (crossover, value, blinder) = utx.crossover();
+        circ.set_fee_crossover(utx.fee(), crossover, *value, *blinder);
+
+        let proof = circ.prove(&PP, &pk).map_err(|e| {
+            Status::invalid_argument(format!(
+                "Failed proving transaction: {}",
+                e
+            ))
+        })?;
+        let proof = proof.to_bytes().to_vec();
+
+        Ok(Response::new(ProverResponse { proof }))
+    }
+}
+
+fn circuit_from_numbers(
+    num_inputs: usize,
+    num_outputs: usize,
+) -> Option<ExecuteCircuit> {
+    match (num_inputs, num_outputs) {
+        (1, 0) => {
+            Some(ExecuteCircuit::ExecuteCircuitOneZero(Default::default()))
+        }
+        (1, 1) => {
+            Some(ExecuteCircuit::ExecuteCircuitOneOne(Default::default()))
+        }
+        (1, 2) => {
+            Some(ExecuteCircuit::ExecuteCircuitOneTwo(Default::default()))
+        }
+        (2, 0) => {
+            Some(ExecuteCircuit::ExecuteCircuitTwoZero(Default::default()))
+        }
+        (2, 1) => {
+            Some(ExecuteCircuit::ExecuteCircuitTwoOne(Default::default()))
+        }
+        (2, 2) => {
+            Some(ExecuteCircuit::ExecuteCircuitTwoTwo(Default::default()))
+        }
+        (3, 0) => {
+            Some(ExecuteCircuit::ExecuteCircuitThreeZero(Default::default()))
+        }
+        (3, 1) => {
+            Some(ExecuteCircuit::ExecuteCircuitThreeOne(Default::default()))
+        }
+        (3, 2) => {
+            Some(ExecuteCircuit::ExecuteCircuitThreeTwo(Default::default()))
+        }
+        (4, 0) => {
+            Some(ExecuteCircuit::ExecuteCircuitFourZero(Default::default()))
+        }
+        (4, 1) => {
+            Some(ExecuteCircuit::ExecuteCircuitFourOne(Default::default()))
+        }
+        (4, 2) => {
+            Some(ExecuteCircuit::ExecuteCircuitFourTwo(Default::default()))
+        }
+        _ => None,
```

Comment:
You might want to reduce the code by removing `ExecuteCircuit` for each arms, the result will be more readable since it will stay in one line each (also, each enum variants repeat `ExecuteCircuit`, and it shouldn't IMVHO, but this is a different topic that has nothing to do with this PR).
```suggestion
    use ExecuteCircuit::*;
    match (num_inputs, num_outputs) {
        (1, 0) => Some(ExecuteCircuitOneZero(Default::default())),
        (1, 1) => Some(ExecuteCircuitOneOne(Default::default())),
        (1, 2) => Some(ExecuteCircuitOneTwo(Default::default())),
        (2, 0) => Some(ExecuteCircuitTwoZero(Default::default())),
        (2, 1) => Some(ExecuteCircuitTwoOne(Default::default())),
        (2, 2) => Some(ExecuteCircuitTwoTwo(Default::default())),
        (3, 0) => Some(ExecuteCircuitThreeZero(Default::default())),
        (3, 1) => Some(ExecuteCircuitThreeOne(Default::default())),
        (3, 2) => Some(ExecuteCircuitThreeTwo(Default::default())),
        (4, 0) => Some(ExecuteCircuitFourZero(Default::default())),
        (4, 1) => Some(ExecuteCircuitFourOne(Default::default())),
        (4, 2) => Some(ExecuteCircuitFourTwo(Default::default())),
        _ => None,
    }
```

Reply:
I wondered how to do this effectively. Also, there should be an issue with around the names of the variants.